tests/test_bytes.rs

```diff
@@ -514,3 +514,103 @@ fn partial_eq_bytesmut() {
     assert!(bytes2 != bytesmut);
     assert!(bytesmut != bytes2);
 }
+
+#[test]
+fn unsplit_basic() {
+    let mut buf = BytesMut::with_capacity(64);
+    buf.extend_from_slice(b"aaabbbcccddd");
+    
+    let splitted = buf.split_off(6);
+    assert_eq!(b"aaabbb", &buf[..]);
+    assert_eq!(b"cccddd", &splitted[..]);
+    
+    buf.unsplit(splitted);
+    assert_eq!(b"aaabbbcccddd", &buf[..]);
+}
+
+#[test]
+fn unsplit_inline_arc() {
+    let mut buf = BytesMut::with_capacity(8); //inline
+    buf.extend_from_slice(b"aaaabbbb");
+    
+    let mut buf2 = BytesMut::with_capacity(64); 
+    buf2.extend_from_slice(b"ccccddddeeee");
+    
+    let _buf3 = buf2.split_off(8); //arc
+
+    buf.unsplit(buf2);
+    assert_eq!(b"aaaabbbbccccdddd", &buf[..]);
+}
+
+#[test]
+fn unsplit_arc_inline() {
+    let mut buf = BytesMut::with_capacity(64); // arc
+    buf.extend_from_slice(b"aaaabbbbeeee");
+
+    let _buf2 = buf.split_off(8); //arc
+
+    let mut buf3 = BytesMut::with_capacity(8); //inline
+    buf3.extend_from_slice(b"ccccdddd");
+
+    buf.unsplit(buf3);
+    assert_eq!(b"aaaabbbbccccdddd", &buf[..]);
+
+}
+
+#[test]
+fn unsplit_both_inline() {
+    let mut buf = BytesMut::with_capacity(16); //inline
+    buf.extend_from_slice(b"aaaabbbbccccdddd");
+    
+    let splitted = buf.split_off(8); // both inline
+    assert_eq!(b"aaaabbbb", &buf[..]);
+    assert_eq!(b"ccccdddd", &splitted[..]);
+
+    buf.unsplit(splitted);
+    assert_eq!(b"aaaabbbbccccdddd", &buf[..]);
+}
+
+
+#[test]
+fn unsplit_arc_different() {
+    let mut buf = BytesMut::with_capacity(64); //arc
+    buf.extend_from_slice(b"aaaabbbbeeee");
+
+    let _buf2 = buf.split_off(8);
+
+    let mut buf3 = BytesMut::with_capacity(64); //arc
+    buf3.extend_from_slice(b"ccccddddeeee");
+
+    let _buf4 = buf3.split_off(8);
+    
+    buf.unsplit(buf3);
+    assert_eq!(b"aaaabbbbccccdddd", &buf[..]);
+}
+
+#[test]
+fn unsplit_arc_non_contiguous() {
+    let mut buf = BytesMut::with_capacity(64); //arc
+    buf.extend_from_slice(b"aaaabbbbeeee");
+
+    let _buf2 = buf.split_off(8);
+
+    let mut buf3 = BytesMut::with_capacity(64); //arc
+    buf3.extend_from_slice(b"eeeeccccdddd");
+
+    let buf4 = buf3.split_off(4);
+    
+    buf.unsplit(buf4);
```

Comment:
In this test `buf` and `buf4` have totally different arcs.  I think it would be a better to test if they shared the same arc, but noncontiguous portions.  Something like:
``` rust
buf.extend_from_slice(b"aaaabbbbeeeeffff");
let buf2 = buf.split_off(8);
let buf4 = buf2.split_off(8);
buf.unsplit(buf4);
```